### PR TITLE
Add support for AWS auto scaling groups

### PR DESCRIPTION
**Pull request by klizhentas**
* Migrate to terraform 0.10.8
* Add resources specs for ASG
* Update user data for masters and nodes
* Add SSM integration
* Add design doc section

### CODE REVIEW

File: scripts/terraform/asg.tf [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+  tag {
+    key                 = "KubernetesCluster"
+    value               = "${var.cluster_name}"
+    propagate_at_launch = true
+  }
+
+  tag {
+    key                 = "Customer"
+    value               = "Mulesoft"
```
Comment (helgi) [inline scripts/terraform/asg.tf]: Do you want this in here?

**Comment by helgi** [comment]
Masters are not in an ASG or Launch Configuration; is that intentional? I know we wouldn't want to scale those up/down but there is more of a question of being able to have AWS replace a faulty master (bad instance or similar), assuming k8s can handle a master leaving and joining back in.

Is the plan to have those be handled differently?

**Comment by klizhentas** [comment]
yes, masters are not in ASG on purpose, because replacing masters is a more involved process. E.g. if machine with masters goes down, we don't want to replace it with another node without cleaning up the state first.

**Comment by helgi** [comment]
That makes sense - How much state does it need repairing tho that can't be done by letting other master known a node has been taken out of the rotation and a new one is coming in? Are there some good docs around for that

**Comment by klizhentas** [comment]
All the automation is in place:

https://gravitational.com/docs/cluster/#removing-a-node

When we remove the master node we reconfigure etcd too.

**Comment by klizhentas** [comment]
Technically, nothing prevents us from creating a special ASG for master nodes, but there will be many edge cases when this whole setup will crash, for example:

Imagine that 2 nodes in ASG out of three will go down, then the ASG will bring 2 more, but the etcd cluster will be out.

**Pull request merged**